src/language/schema/materializer.js

```diff
@@ -0,0 +1,263 @@
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import {
+  TYPE_DEFINITION,
+  INTERFACE_DEFINITION,
+  ENUM_DEFINITION,
+  UNION_DEFINITION,
+  SCALAR_DEFINITION,
+  INPUT_OBJECT_DEFINITION,
+} from './kinds';
+
+import {
+  GraphQLSchema,
+  GraphQLObjectType,
+  GraphQLInterfaceType,
+  GraphQLUnionType,
+  GraphQLScalarType,
+  GraphQLEnumType,
+  GraphQLInputObjectType,
+  GraphQLString,
+  GraphQLInt,
+  GraphQLBoolean,
+  GraphQLID,
+  GraphQLList,
+  GraphQLNonNull,
+} from '../../type';
+
+import {
+  coerceValueAST
+} from '../../executor/values';
+
+import {
+  LIST_TYPE,
+  NON_NULL_TYPE,
+} from '../kinds';
+
+function nullish(obj) {
+  return obj === null || obj === undefined;
+}
+
+/**
+ * Takes the array of N elements constructs an object.
+ * For each elemenet sets obj[keyFnArg(e}] = valueFnArg
+ * If they are not specified they default to the identity function
+ */
+function makeDict(array, keyFnArg, valueFnArg) {
+  var map = {};
+  var keyFn = nullish(keyFnArg) ? t => t : keyFnArg;
+  var valueFn = nullish(valueFnArg) ? t => t : valueFnArg;
+  array.forEach((item) => {
+    map[keyFn(item)] = valueFn(item);
+  });
+  return map;
+}
+
+function buildWrappedType(innerType, inputTypeAST) {
+  if (inputTypeAST.kind === LIST_TYPE) {
+    return new GraphQLList(buildWrappedType(innerType, inputTypeAST.type));
+  }
+  if (inputTypeAST.kind === NON_NULL_TYPE) {
+    return new GraphQLNonNull(buildWrappedType(innerType, inputTypeAST.type));
+  }
+  return innerType;
+}
+
+function getInnerTypeName(typeAST) {
+  if (typeAST.kind === LIST_TYPE || typeAST.kind === NON_NULL_TYPE) {
+    return getInnerTypeName(typeAST.type);
+  }
+  return typeAST.name.value;
+}
+
+/**
+ * This takes the ast of a schema document produced by parseSchema in
+ * src/language/schema/parser.js.
+ *
+ * Given that AST it constructs a GraphQLSchema. As constructed
+ * they are not particularly useful for non-introspection queries
+ * since they have no resolve methods.
+ */
+export function materializeSchema(ast, queryTypeName, mutationTypeName) {
+  if (nullish(ast)) {
+    throw new Error('must pass in ast');
+  }
+  if (nullish(queryTypeName)) {
+    throw new Error('must pass in query type');
+  }
+
+  var astMap = makeDict(ast.definitions, d => d.name.value);
+
+  /**
+   * This generates a function that allows you to produce
+   * type definitions on demand. We produce the function
+   * in order to close over the memoization dictionaries
+   * that need to be retained over multiple functions calls.
+   **/
+  function getTypeDefProducer() {
+
+    var innerTypeMap = {
+      String: GraphQLString,
+      Int: GraphQLInt,
+      Boolean: GraphQLBoolean,
+      ID: GraphQLID,
+    };
+
+    return (typeAST) => {
+      var typeName = getInnerTypeName(typeAST);
+      if (!nullish(innerTypeMap[typeName])) {
+        return buildWrappedType(innerTypeMap[typeName], typeAST);
```

Comment:
I guess we could construct some key scheme based on the ! + [] notation but it seemed unnecessary